### PR TITLE
- add tbb

### PR DESCRIPTION
**Pull request by SSE4**
~~**DRAFT** as neither MSYS make nor nmake work well, need to use mingw-make (not available on CCI yet)~~

Specify library name and version:  **tbb/2019_u9**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/tbb/all/test_package/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,12 @@
+set(CMAKE_CXX_COMPILER_WORKS 1)
```
Comment (Croydon) [inline recipes/tbb/all/test_package/CMakeLists.txt]: What is this about?

File: recipes/tbb/all/test_package/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+PROJECT(TbbExample CXX)
+cmake_minimum_required(VERSION 2.8.12)
```
Comment (Croydon) [inline recipes/tbb/all/test_package/CMakeLists.txt]: Needs to be at top of file

File: recipes/tbb/all/test_package/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
+
+message("LIBS: ${CONAN_LIBS}")
```
Comment (Croydon) [inline recipes/tbb/all/test_package/CMakeLists.txt]:
```suggestion
```

File: recipes/tbb/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+                extra += " stdlib=libc++"
+            extra += " compiler=gcc" if self.settings.compiler == 'gcc' else " compiler=clang"
+
+            extra += " gcc_version={}".format(str(self.settings.compiler.version))
+        make = tools.get_env("CONAN_MAKE_PROGRAM", tools.which("make") or tools.which('mingw32-make'))
+        if not make:
+            raise ConanInvalidConfiguration("This package needs 'make' in the path to build")
+
+        with tools.chdir(self._source_subfolder):
+            # intentionally not using AutoToolsBuildEnvironment for now - it's broken for clang-cl
```
Comment (danimtb) [inline recipes/tbb/all/conanfile.py]: Is there an issue for this in conan-io/conan repo?

File: recipes/tbb/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+        extra = "" if self.settings.os == "Windows" or self.options.shared else "extra_inc=big_iron.inc"
+        if self.settings.arch == "x86":
+            arch = "ia32" 
+        elif self.settings.arch == "x86_64":
+            arch = "intel64"
+        elif self.settings.arch == "armv7":
+            arch = "armv7"
+        elif self.settings.arch == "armv8":
+            arch = "aarch64"
```
Comment (danimtb) [inline recipes/tbb/all/conanfile.py]: You could use a dict for this and it will be cleaner

File: recipes/tbb/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        self.copy(pattern="*%s*.a" % build_type, dst="lib", src=build_folder, keep_path=False)
+        self.copy(pattern="*%s*.dll" % build_type, dst="bin", src=build_folder, keep_path=False)
+        self.copy(pattern="*%s*.dylib" % build_type, dst="lib", src=build_folder, keep_path=False)
+        # Copy also .dlls to lib folder so consumers can link against them directly when using MinGW
+        if self.settings.os == "Windows" and self.settings.compiler == "gcc":
+            self.copy("*%s*.dll" % build_type, dst="lib", src=build_folder, keep_path=False)
+
+        if self.settings.os == "Linux":
+            # leaving the below line in case MacOSX build also produces the same bad libs
+            extension = "dylib" if self.settings.os == "Macos" else "so"
```
Comment (danimtb) [inline recipes/tbb/all/conanfile.py]: The following if is never going to be true as it is under `if self.settings.os == "Linux"` clause

File: recipes/tbb/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+                    self.run("ln -s \"%s\" \"%s\"" %
+                             (fpath, fpath[0:fpath.rfind("." + extension) + len(extension) + 1]))
+
+    def package_info(self):
+        suffix = "_debug" if self.settings.build_type == "Debug" else ""
+        libs = {"tbb": "tbb", "tbbproxy": "tbbmalloc_proxy", "tbbmalloc": "tbbmalloc"}
+        targets = self.get_targets()
+        self.cpp_info.libs = ["{}{}".format(libs[target], suffix) for target in targets]
+        if self.settings.os == "Linux":
+            self.cpp_info.libs.append("pthread")
```
Comment (danimtb) [inline recipes/tbb/all/conanfile.py]: system_libs ?

File: recipes/tbb/all/conanfile.py (hunk 4 above)
Comment (SSE4) [reply recipes/tbb/all/conanfile.py]: yes, there are multiple issues and PRs to support `clang-cl` in conan project already

File: recipes/tbb/all/conanfile.py (hunk 5 above)
Comment (SSE4) [reply recipes/tbb/all/conanfile.py]: uses dict now

File: recipes/tbb/all/conanfile.py (hunk 7 above)
Comment (SSE4) [reply recipes/tbb/all/conanfile.py]: changed to system_libs

File: recipes/tbb/all/conanfile.py (hunk 6 above)
Comment (SSE4) [reply recipes/tbb/all/conanfile.py]: removed condition

File: recipes/tbb/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+                    self.run("ln -s \"%s\" \"%s\"" %
+                             (fpath, fpath[0:fpath.rfind("." + extension) + len(extension) + 1]))
+
+    def package_info(self):
+        suffix = "_debug" if self.settings.build_type == "Debug" else ""
+        libs = {"tbb": "tbb", "tbbproxy": "tbbmalloc_proxy", "tbbmalloc": "tbbmalloc"}
+        targets = self.get_targets()
+        self.cpp_info.libs = ["{}{}".format(libs[target], suffix) for target in targets]
+        if self.settings.os == "Linux":
+            self.cpp_info.system_libs.append("pthread")
```
Comment (uilianries) [inline recipes/tbb/all/conanfile.py]:
```suggestion
            self.cpp_info.system_libs.extend(["dl", "rt", "m", "pthread"])
```

> ldd .conan/data/tbb/2019_u9/_/_/package/7f2fabdc3c40fe90b3e/lib/libtbb.so
>	linux-vdso.so.1 (0x00007fff5db7e000)
>	libdl.so.2 => /usr/lib/libdl.so.2 (0x00007fbed8c14000)
>	librt.so.1 => /usr/lib/librt.so.1 (0x00007fbed8c09000)
>	libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x00007fbed8a1f000)
>	libm.so.6 => /usr/lib/libm.so.6 (0x00007fbed88d9000)
>	libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x00007fbed88bf000)
>	libpthread.so.0 => /usr/lib/libpthread.so.0 (0x00007fbed889d000)
>	libc.so.6 => /usr/lib/libc.so.6 (0x00007fbed86d4000)
>	/usr/lib64/ld-linux-x86-64.so.2 (0x00007fbed8c9d000)

**Pull request merged**